File: Lab_12_MementoPattarn/TextEditor/src/TextEditor.java
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

public class TextEditor implements Serializable
{
    private static final long serialVersionUID = 1L;//The serialVersionUID is a universal version identifier for a Serializable class. Deserialization uses this number to ensure that a loaded class corresponds exactly to a serialized object. If no match is found, then an InvalidClassException is thrown.
    private String content;
    private int cursorPosition;
    private List<String> selections;

    public TextEditor() {
        this.content = "";
        this.cursorPosition = 0;
        this.selections= new ArrayList<>();
    }

    public void setContent(String content) {
        if(content == null) {
            this.content = "";
        } else {
            this.content = content;
        }
    }

    public void setCursorPosition(int position)
    {
        if(position < 0) {
            this.cursorPosition = 0;
        } else if(position > content.length()) {
            this.cursorPosition = content.length();
        } else {
            this.cursorPosition = position;
        }
    }

    public void addSelection(String selection) {
        if(selection !=null && !selection.isEmpty())//if selection is not null and not empty
        {
            selections.add(selection);//add selection to the selections list;
        }

    }

    public void clearSelections() {
        selections.clear();
    }

    public void displayState() {
        System.out.println("Text Editor State");
        System.out.println("Content: " + content);
        System.out.println("Cursor Position: " + cursorPosition);
        System.out.println("Selections (" + selections.size() + "):");
        for (int i = 0; i < selections.size(); i++) {
            System.out.println("  " + (i + 1) + ": " + selections.get(i));
        }
        System.out.println("-------------------------");
    }

    public EditorMemento createMemento() {
        return new EditorMemento(content, cursorPosition, new ArrayList<>(selections));//return a new EditorMemento object with the current content, cursorPosition, and selections
    }

    public void restoreFromMemento(EditorMemento memento)
    {
        if(memento!=null) {
            this.content = memento.getContent();
            this.cursorPosition = memento.getCursorPosition();
            this.selections = new ArrayList<>(memento.getSelections());
        }

    }

    public String getContent() {
        return content;
    }
    public int getCursorPosition() {
        return cursorPosition;
    }

    public List<String> getSelections() {
        return selections;
    }

}

File: TextEditor/MyTextEditor/TextEditor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MyTextEditor
{
    using System;
    using System.Collections.Generic;

    namespace TextEditor
    {
        public class TextEditor
        {
            private string content;
            private int cursorPosition;
            private List<string> selections;

            public TextEditor()
            {
                content = string.Empty;
                cursorPosition = 0;
                selections = new List<string>();
            }

            public void SetContent(string newContent)
            {
                content = newContent;

                if (cursorPosition > content.Length)
                {
                    cursorPosition = content.Length;
                }
            }

            public void SetCursorPosition(int position)
            {
                if (position < 0)
                {
                    cursorPosition = 0;
                }
                else if (position > content.Length)
                {
                    cursorPosition = content.Length;
                }
                else
                {
                    cursorPosition = position;
                }
            }

            public void AddSelection(string selection)
            {
                selections.Add(selection);
            }

            public void ClearSelections()
            {
                selections.Clear();
            }

            public void DisplayState()
            {
                Console.WriteLine("=== Text Editor State ===");
                Console.WriteLine($"Content: {content}");
                Console.WriteLine($"Cursor Position: {cursorPosition}");
                Console.WriteLine("Selections:");

                if (selections.Count == 0)
                {
                    Console.WriteLine("empty");
                }
                else
                {
                    foreach (var selection in selections)
                    {
                        Console.WriteLine(selection);
                    }
                }
                Console.WriteLine("=======================");
            }

        }
    }
}
